CASSANDRA-17056: CEP-17

jacek-lewandowski opened the pull request:
SSTable format API

Summary of the changes:

Format, reader and writer
---------------------------
There are a lot of refactorings around sstable related classes aiming to extract the most generic functionality to the top-level entities and push down implementation-specific stuff to the actual implementation. In Particular, the top-level, implementation agnostic classes/interfaces are SSTableFormat interface, SSTable, SSTableReader, SSTableWriter, IVerifier, and IScrubber. The rest of the codebase has been reviewed for explicit usages of big table format-specific usages of sstable classes and refactored. SSTable, SSTableReader, and SSTableWriter have their builders. Builders make a hierarchy that follows the same inheritance structure as readers and writers. 

There are also partial implementations that add support for some features and may or may not be used by the custom implementations. They include:
- AbstractSSTableFormat - adds an implementation of some initialization methods - in practice, all of the format implementations should extend this class
- SSTableReaderWithFilter - add support for Bloom filter to the reader
- SortedTableWriter - generic implementation for a writer which writes partitions in the default order to the data file, supports Bloom filter and some index of partitions 
- IndexSummarySupport - interface implemented by the readers using index summaries
- KeyCacheSupport - interface implemented by the readers using row key cache

Descriptor
---------------------------
Refactored the Descriptor class so that:
- All paths are created from the base directory File rather than from a String
- All the methods named *filename* producing full paths were made private; their current implementations are returning file names rather than paths (the naming was inconsistent)
- The usages of the `filenameFor` method were refactored to use the `fileFor` method
- The usages of the `fromFilename` method were refactored to use a  `fromFileWithComponent(..., false).left` expression
In essence, the Descriptor class is no longer working on String-based paths. 

Index summaries
---------------------------
Removed the index summary from the generic SSTableReader class and created an interface IndexSummarySupport to be implemented by the readers that need it. Methods in related classes that refer back to the reader were refactored to support just readers of the SSTableReader & IndexSummarySupport type. Therefore, we will no longer need to assume that the generic SSTableReader has anything to do with an index summary. 

A new IndexSummaryComponent class encloses data fields from the index summary file (note that aside from the index summary itself, the file includes the first and last partition of the sstable). The class has been extracted to deal with those fields and have that logic in a single place.

Filter
---------------------------
Refactored IFilter and its serialization - in particular, added the `serialize` method to the IFilter interface and moved loading/saving logic to a separate utility class FilterComponent.
Extracted the SSTableReaderWithFilter abstract reader extending the generic SSTableReader with filter support.
Extracted bloom filter metrics into separate entities allowing to plug them in if the implementation uses a filter.

Cache
---------------------------
Refactored CacheService to support different key-cache values. CacheService now supports arbitrary IRowIndexEntry implementation as a key-cache value. A new version of the auto-saving cache was created ("g") because some information about the type of serialized row index entry needs to be known before it is deserialized (or skipped). Therefore, the SSTableFormat type ordinal number is stored, which is sufficient because the IRowIndexEntry serializer is specific to the sstable format type.
Similarly to the IndexSummarySupport, a new KeyCacheSupport interface has to be implemented to mark the reader as supporting key-cache. It contains the default implementation of several methods the rest of the system relies on when the key-cache is supported.

Other changes
---------------------------
- Fixed disabling chunk cache - enable(boolean) method in ChunkCache does not make any sense - it makes a false impression it can disable chunk cache once enabled, while in fact, it only clears it. Added setFileCacheEnabled to DatabaseDescriptor

- Made WrappingUnfilteredRowIterator an interface

- DataInputStreamPlus extends InputStream - this makes it possible for input stream-based inheritors of DataInputPlus to extend DataInputStreamPlus. It simplifies coding because sometimes we want to get DataInputPlus implementation extending InputStream as an argument.

- Table and keyspace metrics were made pluggable - in particular, added the ability for a certain format to register gauges that are specific only to that format and make no sense for others

- Implemented mmapped region extension for compressed data

- Refactored FileHandle so that it is no longer closable

- Implemented WrappingRebufferer

- Introduced the SSTable.Owner interface to make SSTable implementation not reference higher-level entities directly. SSTable accepts passing null as the owner when there is no owner (like sometimes in offline tools) or passing a mock when needed in tests.

Individual commits
---------------------------

[4a87cd36fe] Fix disabling chunk cache
[c84c75ccf3] Made WrappingUnfilteredRowIterator an interface
[253d2b828e] Add getType to SSTableFormat
[3f169dcc20] Remove getIndexSerializer from SSTableFormat
[05bae1833b] Pull down rowIndexEntrySerializer field
[da675f2809] Moved RowIndexEntry
[673f0c5c39] Reduce usages of RowIndexEntry
[c72538be91] Refactor CacheService to support for different key cache values
[54d33ee656] Minor refactoring of ColumnIndex
[93862df967] Just moved AbstractSSTableIterator to o.a.c.io.sstable.format
[9e4566a1de] Refactored AbstractSSTableIterator
[a4e61e80bb] Extracted IScrubber and IVerifier interfaces
[20f78c7419] Push down implementation of SSTableReader.firstKeyBeyond
[f2c24e5774] Moved SSTableReader.getSampleIndexesForRanges to IndexSummary
[b6c3a6c1ea] Moved SSTableReader.getKeySamples implementation to IndexSummary
[c4b90ebb33] Refactor InstanceTidier so that it is more generic
[918d5a9e74] Refactor dropping page cache
[a52fb4d558] Refactor sstable metrics
[f6d10f930f] NEW (fix up) - DataInputStreamPlus extends InputStream
[8f6a56d972] Getting rid of index summary in SSTableReader
[4a918bf725] Removed direct usages of primary index from SSTableReader
[358fa32602] Refactor KeyIterator so that it is sstable format agnostic
[14c09d89c2] Remove explicit usage of Components outside of format specific classes
[feff14e137] Move clone methods implementation from SSTableReader to BigTableReader
[64e9787b10] Move saveIndexSummary and saveBloomFilter to SSTableReaderBuilder
[ae71fe6ed8] Moved indexSummary field to BigTableReader and made it private
[df9fd8c4b9] Moved ifile field to BigTableReader and made it private
[2be6ea9ecf] Moved static open methods for BigTableReader to the reader factory
[bc0e55ac48] Minor refactoring around IFilter and its serialization
[5b95704beb] Minor refactorings around IndexSummary
[87812335e8] Extracted TOCComponent class to deal with TOC file
[fdad092a6a] Extracted CompressionInfoComponent class
[39b47e388d] Extracted StatsComponent as a helper for elements of SSTable metadata
[cdb55bff47] Fix SSTable.getMinimalKey
[b99c6d5805] Refactor FileHandle so that it is no longer closable
[77b7f7ace5] Implement WrappingRebufferer
[b6868914dd] Add progressPercentage to ProgressInfo
[7fd4956e5b] Moved copy/rename/hardLink methods from SSTableWriter to SSTable
[1ccc6bf148] Create generic SSTableBuilder and IOOptions
[da58a81102] Refactor SSTableReaderBuilder
[4501ddba1c] Refactor ColumnIndex
[d4f9e1a64b] Extracted non-big-table-specific functionality from BigTableWriter to SortedTableWriter
[379525d01e] Refactor BigTableZeroCopyWriter to SSTableZeroCopyWriter as it is not specific to big format
[8ac37f83bc] Extract EmptySSTableScanner out from BigTableScanner
[ee6673f1cf] Implement SSTableWriterBuilder
[bb26629235] Refactor opening early / final
[a327595015] Refactored SSTableWriter factory
[16ffd7334b] Extract non-big-format-specific logic from scrubber and verifier
[75e02db6af] Allow to specify the default SSTableFormat via system property
[a7b9d0d628] Small fixes around streaming
[407f977c36] Move guard collection size
[0529e57d2f] Remove explicit references to big format
[61509963ec] Unclassified minor changes
[da28d1af3a] Replaced getCreationTimeFor(Component) with getDataCreationTime()
[e99c834de6] !!! Reformatting
[882b7baa5a] Rename SSTableReader.maybePresent and fix its redundant usages
[b70c983bea] Implement mmapped region extension for compressed data
[d7ff3970de] Introduce SSTable.Owner interface
[e9feb9c462] Replaced getCreationTimeFor(Component) with getDataCreationTime()
[ee8082fb07] Created SSTableFormat.deleteOrphanedComponents
[e62950fd3d] Refactor metrics further
[cefa5b3814] Extract key cache support into separate entity
[dd55101ca1] Extracted SSTableReaderWithFilter
[510b651824] Implement customizable component types
[2be512d9fa] Pluggable SSTableFormat by making SSTableFormat.Type not an enum
[670836b55d] Refactor CRC and digest validators
[00c91103bc] Extract delete method to delete SSTables and purge row cache entries
[0819dc9fc2] Extracted trySkipFileCacheBefore(key) to SSTableReader
[732f841750] Added missing overrides in ForwardingSSTableReader
[db623218fd] Update DatabaseDescriptorRefTest
[c018c468e5] Cleanup
[eafc836242] Add @SuppressWarnings("resource") where needed
[3b7c911dd6] Documentation

patch by Jacek Lewandowski, reviewed by Branimir Lambov for CASSANDRA-17056

Co-authored-by: @jacek-lewandowski 
Co-authored-by: @blambov 

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/CASSANDRA-17056)

blambov commented on src/java/org/apache/cassandra/utils/Throwables.java:
```diff
@@ -189,13 +189,19 @@ public static Throwable perform(Throwable accumulate, String filePath, FileOpTyp
             try
             {
-                closeable.close();
+                if (closeable != null)
```
I'm not sure we want to relax this. Usually `FileUtils.close` is used to close things that may include nulls.

jacek-lewandowski replied:
This is a new method. `FileUtils.close` just throws while here we are adding suppressed exceptions. Actually this works very similar to try..catch with resources, where only non-null resources are closed and no NPE is thrown. If you think I should move it to `FileUtils` I'm ok with that.

blambov replied:
I guess the comment applies to the method below then, but if we don't change it there it would be unexpected to skip here &mdash; unless we make the naming very clear about that difference.

jacek-lewandowski replied:
9ffff01f36d271686bac16b932b57c7d918a355d as well

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/db/compaction/Scrubber.java:
```diff
@@ -742,6 +750,12 @@ public FixNegativeLocalDeletionTimeIterator(UnfilteredRowIterator iterator, Outp
+        {
+            return iterator;
+        }
+
```
Should we not take advantage of the methods implemented by the wrapping interface, i.e. remove the implementations below?

jacek-lewandowski replied:
yes, sorry, oversight

jacek-lewandowski replied:
it was actually applied in some of the next steps, but it should be in this commit

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/db/compaction/Scrubber.java:
```diff
@@ -564,20 +565,27 @@ public ScrubResult(Scrubber scrubber)
      * For more details, refer to CASSANDRA-12144.
      */
-    private static class RowMergingSSTableIterator extends WrappingUnfilteredRowIterator
+    private static class RowMergingSSTableIterator extends UnmodifiableIterator<Unfiltered> implements WrappingUnfilteredRowIterator
```
Could we not implement `remove` in the interface so that `extends UnmodifiableIterator` is not necessary here and elsewhere?

jacek-lewandowski replied:
So `remove` is implemented in Java's `Iterator` in the same way as in `UnmodifiableIterator` - did you mean that?

blambov replied:
Yes, we can supply a default implementation in the interface just like it is done in `UnmodifiableIterator`, and not have to descend from the latter.

jacek-lewandowski replied:
We do not have to supply anything, the Java's original implementation of `Iterator.remove` is:

```
    default void remove() {
        throw new UnsupportedOperationException("remove");
    }
```

`UnmodifiableIterator` just makes it final

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/service/CacheService.java:
```diff
@@ -439,13 +440,14 @@ public void serialize(KeyCacheKey key, DataOutputPlus out, ColumnFamilyStore cfs
+            // format type id is stored so that in case there is no sstable for the key we can figure out which
+            // serializer (of which sstable format) was used and thus as can use the right implemnentation to skip
+            // the unmatched entry
+            out.writeByte(key.desc.formatType.ordinal());
```
Would it be simpler to store a (varint) length here?

jacek-lewandowski replied:
I know it would be more elegant and I thought about that, but on the other hand, we would have to have `serializedSize` at hand or buffer the whole entry. Serialized size is not available without going through all the columns (for indexed row entry) - I don't know if it is worth the effort. That's why I stored format ordinal

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableFormat.java:
```diff
@@ -36,6 +36,9 @@
     SSTableWriter.Factory getWriterFactory();
     SSTableReader.Factory getReaderFactory();
 
+    boolean isKeyCacheSupported();
```
I would call this `cachesKeys` (if the sstable impl takes care of doing the caching calls) or `cacheKeys`/`shouldCacheKeys` (if other code is doing it) -- the alternate implementation can support cached keys, but does not need to.

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/AbstractSSTableIterator.java:
```diff
@@ -382,31 +406,174 @@ public Unfiltered next()
+        public String toString()
+        {
+            return file != null ? file.toString() : "null";
+        }    }
```
Nit: formatting needs fixing.

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableFormat.java:
```diff
@@ -36,6 +39,8 @@
     SSTableWriter.Factory getWriterFactory();
     SSTableReader.Factory getReaderFactory();
 
+    Set<Component> supportedComponents();
```
As below, "supported" is not the right term here. Maybe "included"? (I believe we also have "required" for the mandatory ones?)

jacek-lewandowski replied:
I don't know, `supported` seemed to most relevant to me as it is intended to return all the components that can be loaded/used/supported by that format.

jacek-lewandowski replied:
OTOH, it can be also, something like `written`, or `produced` as this is also all the components that can be created when sstable is written.

jacek-lewandowski replied:
maybe will change this to just `all`

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/schema/Schema.java:
```diff
@@ -224,6 +225,15 @@ public ColumnFamilyStore getColumnFamilyStoreInstance(TableId id)
+    public ColumnFamilyStore getColumnFamilyStoreInstance(TableMetadata metadata)
+    {
+        ColumnFamilyStore cfs = getColumnFamilyStoreInstance(metadata.id);
+        if (cfs == null || !metadata.isIndex())
```
Shouldn't this be `!=`?

jacek-lewandowski replied:
no, if it is null we want to return

blambov commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableReader.java:
```diff
@@ -374,4 +380,20 @@ public ClusteringPrefix<?> getLowerBoundPrefixFromCache(DecoratedKey partitionKe
+    @Override
+    public IScrubber getScrubber(LifecycleTransaction transaction, OutputHandler outputHandler, IScrubber.Options options)
+    {
+        ColumnFamilyStore cfs = Schema.instance.getColumnFamilyStoreInstance(metadata());
```
Could we do this without referencing the higher-level CFS and Schema concepts in the sstable reader implementation?

Perhaps implement it in the format and pass the CFS as an argument?

jacek-lewandowski replied:
I wanted to avoid requiring CFS as an argument but you are right.

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/schema/Schema.java:
```diff
@@ -224,6 +225,15 @@ public ColumnFamilyStore getColumnFamilyStoreInstance(TableId id)
                : null;
     }
 
+    public ColumnFamilyStore getColumnFamilyStoreInstance(TableMetadata metadata)
```
The behaviour of this is different from id lookup. Could we give it a different name?

jacek-lewandowski replied:
actually, after applying your fixes this method is not needed any longer; so I will just add some comment to the original method

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -808,32 +809,12 @@ public void runOnClose(final Runnable runOnClose)
-            if (runFirst == null)
-                return runSecond;
-            return new AndThen(runFirst, runSecond);
+            tidy.runOnClose = () -> {
```
AFAIR `AndThen` here is necessary because a lambda retains a reference to the reader, which we definitely don't want (see comment on top of the method).

If you revert this change, please add a "do not replace with a lambda..." comment in front of `AndThen` itself.

jacek-lewandowski replied:
I must disagree. To my knowledge lambda does not include a reference to the enclosing instance unless explicitly `this` or some member is used. I've verified that, also see this answer https://stackoverflow.com/a/35636556

blambov replied:
[This](https://issues.apache.org/jira/browse/CASSANDRA-9423?focusedCommentId=14602771&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14602771) is the original problem stemming from this. AFAIR Stefania later fought the same problem with lambdas (where `this` was included in the closure even when it was not required), but I may be wrong, or things may have changed.

If you have verified that leak detection works correctly for all supported JDK versions, I'll be very happy to accept that this is one less problem we have to worry about.

jacek-lewandowski replied:
added test cases in `RefCountedTest`

blambov replied:
Great.

Could you add comments explaining that lambdas, if certain that they don't reference `this`, are okay?

As a nit, `existing` can be checked immediately, and `tidy.runOnClose` can be set to the argument directly if it's null.

blambov replied:
The comment looks good, and thank you for adding the test.

To make sure we are checking the `runOnClose` lambda (which is now only constructed on the second call), let's call it twice in the test.

jacek-lewandowski replied:
fee2c251d370fd61ca436bc16169c1d1774ec01c

blambov replied:
The new tests look great. Thank you for adding them.

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1973,17 +1970,36 @@ public void run()
+                        exceptions = ex;
+                    }
+
+                    Throwable closeExceptions = Throwables.close(null, closeables);
```
Isn't `FileUtils.close` a better match for what is done here?

jacek-lewandowski replied:
I'm not sure TBH, need to get the exceptions reported and log them, add them to the `exceptions`. With `FileUtils.close` I'd have to catch the exception.

blambov replied:
It's practically the same pattern, since you do have a check for this concrete part throwing exceptions.

The reason for asking is to avoid relaxing `Throwables.close` -- we may lose an error message we care about for something that should not be null.

Alternatively, we can add a `closeIgnoringNulls` method to `Throwables`.

jacek-lewandowski replied:
I like that alternative idea

jacek-lewandowski replied:
9ffff01f36d271686bac16b932b57c7d918a355d

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -907,43 +907,18 @@ public SSTableReader cloneWithNewStart(DecoratedKey newStart, final Runnable run
+                    handleAndPositions.put(dfile, getPosition(newStart, Operator.EQ));
+                if (ifile != null)
+                    handleAndPositions.put(ifile, getIndexScanPosition(newStart));
+                runOnClose(() -> handleAndPositions.forEach(FileHandle::dropPageCache));
```
The same comment as in `runOnClose` applies -- this lambda retains a reference to the reader.

jacek-lewandowski replied:
I hope this will convince you:

```java
        class SomeUnserializableClass {
            public String a = "x";
            public byte[] serialize(String s) {
                return SerializationUtils.serialize((Supplier<String> & Serializable) () -> {
                    return s + "#" + s + a;
                });
            }
        }

        SomeUnserializableClass serializer = new SomeUnserializableClass();
        byte[] serialized = serializer.serialize("abc");
        Supplier<String> deseraizlied = SerializationUtils.deserialize(serialized);
        assertEquals("abc#abc", deseraizlied.get());
```

this test indeed does not work because of `NotSerializableException` - we try to access `a` in lambda, which requires serialization of `SomeUnserializableClass`. 

When `a` is removed, so that we just have:

```java
return s + "#" + s;
```

in lambda, test tests passes.

blambov commented on src/java/org/apache/cassandra/io/sstable/format/big/BigFormat.java:
```diff
@@ -125,6 +130,24 @@ public AbstractRowIndexEntry.KeyCacheValueSerializer<BigTableReader, RowIndexEnt
+    @Override
+    public BigTableReader cast(SSTableReader sstr)
+    {
+        return sstr == null ? null : (BigTableReader) sstr;
```
There is no need for the null check here. Nulls can be cast to any type.

blambov replied:
looks good

blambov reviewed (commented):
First batch of comments (up to commit "Getting rid of index summary part 1").

blambov commented:
One other thing I think we should include in this is breaking the dependency loop between `SSTableReader` and `ColumnFamilyStore`. `SSTableReader` is a low-level class and should not know how its owner works, and even less about how to get hold of its owner via `Schema`.

If necessary, we can introduce an `SSTableReader.Owner` interface (similar to `Memtable.Owner`) to expose only functionality it needs.

blambov commented on src/java/org/apache/cassandra/db/compaction/CompactionManager.java:
```diff
@@ -1890,22 +1890,17 @@ public void run()
     }
 
-    public List<SSTableReader> runIndexSummaryRedistribution(IndexSummaryRedistribution redistribution) throws IOException
+    public <T, E extends Throwable> T runWithActiveCompactions(Holder holder, ThrowingSupplier<T, E> callable) throws E
```
I don't understand this method's naming -- it is running the callable wrapped as an active compaction with the given info holder?

Maybe `runAsActiveCompaction(CompactionInfo.Holder activeCompactionInfo, ThrowingSupplier callable)`? 
Or `<T, E extends Throwable, R extends CompactionInfo.Holder & ThrowingSupplier<T, E>> T runAsActiveCompaction(R callable) throws E` and use one argument for both operation info and callable to make it explicit that the info is for the operation that the callable does?

jacek-lewandowski replied:
ok

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1033,6 +1012,14 @@ protected void closeInternalComponent(AutoCloseable closeable)
+     * {@link #closeInternalComponent(AutoCloseable)} for each such component. Leaving the implementation empty is
+     * valid, but may impact memory usage.
+     */
+    public abstract void releaseComponents();
```
Nit: This should include some adjective for the type of components released, e.g. `releaseInMemoryComponents`.

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/IndexSummary.java:
```diff
@@ -352,6 +353,37 @@ public void remove()
         };
     }
 
+    public long getScanPositionFromBinarySearch(PartitionPosition key)
```
`FromBinarySearch` does not make much sense here. `getScanPosition(PartitionPosition)` would be better.

blambov replied:
looks good

blambov commented on test/unit/org/apache/cassandra/io/sstable/IndexSummaryTest.java:
```diff
@@ -388,10 +387,10 @@ private void testPosition(IndexSummary original, IndexSummary downsampled, List<
         for (DecoratedKey key : keys)
         {
-            long orig = SSTableReader.getIndexScanPositionFromBinarySearchResult(original.binarySearch(key), original);
+            long orig = original.getScanPositionFromBinarySearchResult(original.binarySearch(key));
```
This can be `original.getScanPosition(key)`.

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/index/sasi/SASIIndexBuilder.java:
```diff
@@ -82,46 +82,46 @@ public void build()
-                            long position = sstable.getPosition(key, SSTableReader.Operator.EQ);
-                            dataFile.seek(position);
-                            ByteBufferUtil.readWithShortLength(dataFile); // key
+                        long position = sstable.getPosition(key, SSTableReader.Operator.EQ);
```
Isn't this `keys.dataPosition()`?

jacek-lewandowski replied:
🤦🏻‍♂️ 
indeed

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/index/sasi/SASIIndexBuilder.java:
```diff
@@ -56,22 +55,23 @@ class SASIIndexBuilder extends SecondaryIndexBuilder
+        long totalBytesToProcess = 0;
         for (SSTableReader sstable : sstables.keySet())
-            totalIndexBytes += getPrimaryIndexLength(sstable);
+            totalBytesToProcess += sstable.uncompressedLength();
```
We probably need a documentation change for this (progress changed from index bytes to data bytes).

jacek-lewandowski replied:
I don't know where, the existing usages of `CompactionInfo` look like they already assumed the data size rather than the index size.

blambov replied:
I can't find any doc online that states this, so I don't think we need to do anything here.

blambov commented on src/java/org/apache/cassandra/io/sstable/Descriptor.java:
```diff
@@ -171,6 +172,18 @@ public List<File> getTemporaryFiles()
         return ret;
     }
 
+    public Set<Component> getComponents(Set<Component> alwaysAdd, Set<Component> optional)
```
Nit: alwaysAdd -> mandatory?

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1796,9 +1654,33 @@ public static void resetTidying()
     }
 
-    public static abstract class Factory
+    public interface Factory<T extends SSTableReader>
```
Does this type parameter provide any benefit? The user shouldn't know the actual type and will generally keep the generic `SSTableReader` reference.

jacek-lewandowski replied:
yes, at this point it does not make sense so I'll remove from this commit

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/FilterComponent.java:
```diff
@@ -0,0 +1,93 @@
+        {
+            filter.serialize(stream, descriptor.version.hasOldBfFormat());
+            stream.flush();
+            stream.sync(); // is it needed if we close the file right after that?
```
This call maps to `FileChannel.force`, which ensures that the data is flushed out to disk (i.e. persistent) before returning. Closing does not do that (it does do the flushing, but we want that to happen before the sync).

jacek-lewandowski replied:
thanks

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/FilterComponent.java:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    public static void saveOrDeleteCorrupted(Descriptor descriptor, IFilter filter) throws IOException
```
The name is a little misleading, because it implies one or the other while we do one or both. Maybe `saveWithDeleteOnFailure` or something similar?

jacek-lewandowski replied:
I'm bad at naming, will change

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/utils/BloomFilterSerializer.java:
```diff
@@ -17,47 +17,66 @@
 import org.apache.cassandra.utils.obs.OffHeapBitSet;
 
-public final class BloomFilterSerializer
+public final class BloomFilterSerializer<I extends InputStream & DataInputPlus, O extends OutputStream & DataOutputPlus> implements IGenericSerializer<BloomFilter, I, O>
```
Is there any benefit from using these types rather than `DataInput/OutputStreamPlus`?

jacek-lewandowski replied:
long story short, because `FileInputStreamPlus` is not a subclass of `DataInputStreamPlus`

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReaderBuilder.java:
```diff
@@ -301,11 +257,11 @@ public SSTableReader build()
-                                                                                                                .withChunkCache(ChunkCache.instance))
+                                               .mmapped(DatabaseDescriptor.getIndexAccessMode() == Config.DiskAccessMode.mmap)
+                                               .withChunkCache(ChunkCache.instance);
+                 FileHandle.Builder dbuilder = new FileHandle.Builder(descriptor.filenameFor(Component.DATA)).compressed(compression)
```
Nit: perhaps move to new line and align with `new` as above?

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReaderBuilder.java:
```diff
@@ -439,12 +396,12 @@ void load(boolean recreateBloomFilter,
+            try (FileHandle.Builder ibuilder = new FileHandle.Builder(descriptor.filenameFor(Component.PRIMARY_INDEX))
+                                               .mmapped(DatabaseDescriptor.getIndexAccessMode() == Config.DiskAccessMode.mmap)
+                                               .withChunkCache(ChunkCache.instance);
+                 FileHandle.Builder dbuilder = new FileHandle.Builder(descriptor.filenameFor(Component.DATA)).compressed(components.contains(Component.COMPRESSION_INFO))
```
Nit: perhaps move to new line and align with new as above?

blambov commented on src/java/org/apache/cassandra/io/sstable/format/big/IndexSummaryComponent.java:
```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    public void saveOrDeleteCorrupted(Descriptor descriptor) throws IOException
```
As for bloom filter, change to `WithDeleteIfFailed` or similar.

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/Descriptor.java:
```diff
@@ -423,6 +425,19 @@ public boolean isCompatible()
+        Set<Component> components = Sets.newHashSetWithExpectedSize(knownTypes.size());
+        for (Component.Type componentType : knownTypes)
+        {
+            Component component = new Component(componentType);
```
Any reason to not make these singletons?

jacek-lewandowski replied:
I've just copied this method, but I suppose it is because of secondary index component, which isn't a singleton. Though, I don't think this method would ever find any secondary index component with such implementation.

jacek-lewandowski replied:
I'll add a singletons set to the `Component`

blambov replied:
Nit: Move the declaration of `singletons` after the components -- this avoids making people wonder how it's populated.

jacek-lewandowski replied:
I don't think I can do that because in that case when `new Component` was called, the `singletons` field would not be initialized yet.

Given your other comments, we will probably go back to `Component` class when making sstable format really pluggable

blambov commented on src/java/org/apache/cassandra/io/sstable/SSTable.java:
```diff
@@ -155,9 +155,9 @@ public DecoratedKey decorateKey(ByteBuffer key)
      */
     public static DecoratedKey getMinimalKey(DecoratedKey key)
```
Nit: could this be renamed? E.g. to `onHeapMinimized`? Or something in the vein of `retainable` like the clusterings?

jacek-lewandowski replied:
actually, no need this method to be in `SSTable`, I'll move it to `DecoratedKey`

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReaderBuilder.java:
```diff
@@ -395,12 +393,13 @@ void load(boolean recreateBloomFilter,
-                                                                                                             .withChunkCache(ChunkCache.instance))
+            try
             {
+                FileHandle.Builder ibuilder = new FileHandle.Builder(descriptor.fileFor(Component.PRIMARY_INDEX))
```
Is everything that can go wrong / throw an exception covered in the `complete` block? Can we leave mmap regions or memory (e.g. compression offsets) unreleased?

blambov replied:
`FileHandleBuilder` being closeable always annoyed me; I do prefer to make it a trivial builder that cannot fail or that doesn't manage any state. However, we will have to manage that state explicitly in its users now.

jacek-lewandowski replied:
hmm... as I'm looking at this now, I think it needs some fixes - channel proxy is not closed in case of a failure. I also missed `CompressionMetadata`... it is  not closeable. I think maybe could implement `SharedCloseable`?

blambov replied:
Can we also remove `lastRegions` for now, as it can be leaked or released while still in use?

We need to separately figure out how to reuse memmappings for the index and data readers -- perhaps give the builder some `ReuseableRegions` to use, to be owned and closed by the writer.

jacek-lewandowski replied:
I can do that, though I'm pretty convenient about safety of the current approach; On the other hand, I must agree with you that it will be more consistent and elegant to move the management of that outside.

jacek-lewandowski replied:
What do you think about instead of managing it in a builder, creating a method in `FileHandle`, something like `duplicateWithExtendedLength(CompressionMetadata, long)`, which retrieves regions from the current rebufferer if applicable (or we could keep mmapped  regions as a field in `FileHandle`)?

This way, it should be super clear - we create a copy of regions when we really copy a file handle

blambov replied:
The duplication idea sounds good to me.

blambov reviewed (commented):
Second batch of comments (up to "Refactor FileHandle").

blambov commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableWriter.java:
```diff
@@ -361,22 +357,27 @@ public SSTableReader openEarly()
             ifile = iwriter.builder.bufferSize(indexBufferSize).complete(boundary.indexLength);
-            if (compression)
-                dbuilder.withCompressionMetadata(((CompressedSequentialWriter) dataFile).open(boundary.dataLength));
+            try (FileHandle.Builder dbuilder = new FileHandle.Builder(descriptor.filenameFor(Component.DATA)))
```
Doesn't this disable the only application of the `regions` reuse in the file handle builder?

jacek-lewandowski replied:
Indeed, it will be fixed in some later commit

blambov commented on src/java/org/apache/cassandra/io/sstable/format/FilterComponent.java:
```diff
@@ -90,4 +97,63 @@ public static void saveOrDeleteCorrupted(Descriptor descriptor, IFilter filter)
+     * Optionally loads a Bloom filter. If the filter is not needed (FP chance is neglectable), it sets
+     * {@link AlwaysPresentFilter} as a filter in the builder. If the filter is expected to be recreated for various
+     * reasons, it leaves it {@code null} (unchanged). Otherwise, it attempts to load the filter, and if it succeeds,
+     * it is set in the builder. If a filter fails to load, it is left {@code null} (unchanged) meaning that it should
```
This comment needs adjusting &mdash; the filter is returned rather than set in the builder.

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/FilterComponent.java:
```diff
@@ -90,4 +97,63 @@ public static void saveOrDeleteCorrupted(Descriptor descriptor, IFilter filter)
+        return !(Math.abs(1 - fpChance) <= filterFPChanceTolerance);
+    }
+
+    static boolean isFPChanceDiffNeglectable(double fpChance1, double fpChance2)
```
Nit: We should use "negligible" instead of "neglectable" (archaic form).

blambov replied:
Nit: there's one "neglectable" left in the comment.

blambov replied:
... and also in the trace message.

jacek-lewandowski replied:
c46630091965b87b2ac4b1f2d4d2e25bfa3f3683

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableFormat.java:
```diff
@@ -103,4 +105,11 @@ public static Type validate(String name)
+    {
+        SSTableReaderBuilder<R, B> builder(Descriptor descriptor);
+
+        SSTableReaderLoadingBuilder<R, B> builder(Descriptor descriptor, TableMetadataRef tableMetadataRef, Set<Component> components);
```
The difference between these two is very unclear. Could you add comments or rename?

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableFlushObserver.java:
```diff
@@ -57,4 +58,6 @@
      */
     void complete();
+
+    void staticRow(Row staticRow);
```
This is best placed after `startPartition` and should have JavaDoc.

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableWriter.java:
```diff
@@ -153,183 +153,228 @@ private CompressionParams compressionFor(final OperationType opType)
-            if (!row.isEmpty())
-                cellCount += Rows.collectStats(row, collector);
-            return row;
+            assert unfiltered.isRangeTombstoneMarker();
```
Nit: I would remove this -- the cast will check anyway, and this is a rather hot path.

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java:
```diff
@@ -0,0 +1,294 @@
+     * @throws FSWriteError if write to the dataFile fails
+     */
+    @Override
+    public final RIE append(UnfilteredRowIterator partition)
```
There's only one use of the returned index entry, and it does not require the precise type. Can we make this return `AbstractRowIndexEntry` and do away with the parameter?

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/SSTableRewriter.java:
```diff
@@ -43,11 +43,11 @@
- * On abort we restore the original lower bounds to the existing readers and delete any temporary files we had in progress,
- * but leave any hard-links in place for the readers we opened to cleanup when they're finished as we would had we finished
+ * On abort, we restore the original lower bounds to the existing readers and delete any temporary files we had in progress,
+ * but leave any hard-links in place for the readers we opened to clean-up when they're finished as we would have finished
```
This does not have the same meaning, "had we" in this case stands for "if we had"

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableWriterBuilder.java:
```diff
@@ -0,0 +1,91 @@
+        try
+        {
+            rowIndexEntrySerializer = new RowIndexEntry.Serializer(descriptor.version, getSerializationHeader());
+            dataWriter = DataComponent.buildWriter(descriptor,
```
Setting these fields here and grabbing them via the getter is somewhat unsafe; e.g. if the builder is reused we could mix instances up.

Can't we just pass anything we construct as arguments to the `BigTableWriter` constructor?

jacek-lewandowski replied:
I did it that way in order to avoid long arguments lists on constructor. Also having say, data file handle and partition writer passed as separate fields in the constructor would theoretically let pass a handle to a different file than the provided partition writer operates on. As far as I understand, those two parameters are inherently related and passing them always together (in a single builder) is more consistent to me.

The resource fields are cleaned in finally block, so the builder can be safely reused.

blambov replied:
Well, usually builders are so that end users don't use long argument lists; the internals can, and in this case it makes better sense that they do. If necessary, the builder can be made a static inner class so that the relevant constructor is private.

jacek-lewandowski replied:
We cannot do a private constructor with this approach because we have a hierarchy - some components are consumed by the constructor of `SortedTableWriter` and the others by `BigTableWriter`. Obviously such a constructor in `SortedTableWriter` would have to be available for its descendants. 

I understand your point, though. I know it is not a perfect approach but I also don't like having different approaches for writer and reader. I think I can try to meet your worries by doing it a bit differently - while not retiring passing stuff in a builder, I can refactor the initialization a bit: f932e370396a287c4293c6698dd9590d5258b72b

blambov replied:
This is overdoing it -- making sure that `buildInternal` cannot be called more than once would suffice.

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableWriter.java:
```diff
@@ -217,7 +218,7 @@ public void setOpenResult(boolean openResult)
     /**
      * Open the resultant SSTableReader before it has been fully written
```
Doc should explain the consumer.

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:
```diff
@@ -331,7 +331,9 @@
-    COMMIT_LOG_REPLAY_LIST("cassandra.replayList", null)
+    COMMIT_LOG_REPLAY_LIST("cassandra.replayList", null),
+
+    SSTABLE_FORMAT_DEFAULT("cassandra.sstable.format.default", "BIG")
```
Nit: I personally prefer the default to be given in context, i.e. in the format class (which will make the new `getEnum` unnecessary).

jacek-lewandowski replied:
heh... I've tried that, it creates some weird static initialization (dis)order problems

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SortedTableWriter.java:
```diff
@@ -0,0 +1,421 @@
+    {
+        int dataBufferSize = ioOptions.diskOptimizationStrategy.bufferSize(statsMetadata.estimatedPartitionSize.percentile(ioOptions.diskOptimizationEstimatePercentile));
+
+        FileHandle dataFile = dataFileBuilder.mmapped(ioOptions.defaultDiskAccessMode == Config.DiskAccessMode.mmap)
```
This should be `mmapped(ioOptions.defaultDiskAccessMode)`, after which the `mmapped(boolean)` version can be removed.

blambov replied:
looks good

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1983,39 +1380,36 @@ public void addTo(Ref.IdentityCollection identities)
+        });
     }
 
     public boolean maybePresent(DecoratedKey key)
```
Could this be renamed to `mayBePresent` or better still `couldContain`? The current spelling implies "present" is used as a verb, i.e. that this is a function that changes some state.

jacek-lewandowski replied:
Sure, but as I'm looking at this method, I'm not sure what's the contract of it? depending on the filter we return whether it is present or may be present. Though, the usages suggest that we rather expect it to return a definite answer. Is this method really needed in that form?

I'd rather implement it this way:

```java
return key.compareTo(first) >= 0 && key.compareTo(last) <= 0 && bf.isPresent(key) && getPosition(key, EQ) >= 0
```

wdyt?

blambov replied:
This is the method by which we check if a tombstone needs to be preserved because data may be present in some older sstable. It should return true if the partition can be contained in the sstable, and false only if it definitely isn't contained, and needs to be as fast as possible (it is called for every partition with a tombstone for every older sstable during compaction).

So if outside of range (which should already be checked btw, it may be worthwhile to see if this check needs to be here) or bloom filter is negative, we _can_ return false, and returning true otherwise is fine (acceptable false positive frequency) as long as we do have a bloom filter. If we don't, we need to do more work or tombstones will never be removed.

jacek-lewandowski replied:
You are right, it is not needed to check. Though, I don't like the fact that this method does implicit assumption for certain usages.

Btw. thanks you pointing me to check the actual usages. It turned out that there were some redundant calls.


8787244fbac13b8b9dee907d8cda0964bf0d047e

blambov replied:
looks good

blambov reviewed (commented):
The comments below complete my initial pass.

I have a couple of general questions:
- The sstable format is still an enum, i.e. cannot be extended by custom linked implementations. This was meant to prevent pluggability and IMHO is against the spirit of the CEP. Perhaps we should let users implement and point to their own `SSTableFormat` implementation class?
- Can we move the potentially but not necessarily shared functionality down into abstract subclasses of e.g. `SSTableReader`? Specifically bloom filter and key caching support?
- Are all the generic parameters helpful? They create clutter and slow down coding, especially on classes where the consumer does not know the end type, e.g. `SSTableBuilder`, `SSTableWriter`. Perhaps in some cases we can define a non-generic interface to avoid the pollution of the generic abstract class?

blambov commented on src/java/org/apache/cassandra/io/sstable/Component.java:
```diff
@@ -35,43 +41,49 @@
+        public static final Type CUSTOM = new Type("CUSTOM",null, name -> new Component(Type.CUSTOM, name));
 
-        final String repr;
+        public static final ImmutableSet<Type> all = ImmutableSet.copyOf(singletonsCollector);
```
This being final and immutable means we can't register a new type, doesn't it?

E.g. the `BigComponentType`s probably aren't here.

Perhaps wrap it an `UnmodifiableSet` like the components?

jacek-lewandowski replied:
yes, more work is in some next commit, just wanted you to take an overview of the direction

jacek-lewandowski replied:
https://github.com/apache/cassandra/pull/2064/commits/5689be8c3a242074a29663be46a06c820020416a

blambov commented on src/java/org/apache/cassandra/io/util/MmappedRegions.java:
```diff
@@ -140,6 +147,25 @@ public void extend(long length)
+    {
+        assert !isCopy() : "Copies cannot be extended";
+
+        if (compressionMetadata.dataLength <= state.length)
```
Nit: there should be a comment why `dataLength` is used here and `compressedFileLength` below. (If I'm not mistaken the data length might be truncated to shorter than the end of the last chunk, while we do want to map the whole chunk to be able to read it).

jacek-lewandowski replied:
I'll add a comment

jacek-lewandowski replied:
Actually, there is no comment to add. It is simply a bug. Not a harmful one, but a bug.

jacek-lewandowski replied:
https://github.com/apache/cassandra/pull/2064/commits/5022c14d3583666f936c3467b4594bade7fbd360

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java:
```diff
@@ -1415,7 +1418,7 @@ public boolean mayContainAssumingKeyIsInRange(DecoratedKey key)
     {
         private final Descriptor descriptor;
-        private final TableId tableId;
+        private final Owner owner;
```
Does it make sense for this to be a weak reference? Can the CFS be collected before the sstables?

jacek-lewandowski replied:
I don't know TBH. But since the owner is passed explicitly when the sstable instance is built, I would find it quite weird that sstable could lost such reference. Especially that the building code doesn't necessarily need to pass an owner, it is allowed to pass null. This is just my gut feeling, and I'm not strongly opposed to change that.

blambov replied:
I guess I ended up asking the wrong question, I was wondering if it's okay to use the CFS after it has been released.

But the CFS is not reference counted and explicitly released, so this should be fine. And I wouldn't worry about holding on to that reference a little longer than strictly necessary.

blambov commented on src/java/org/apache/cassandra/io/sstable/Component.java:
```diff
@@ -41,55 +39,76 @@
+            this.formatClass = formatClass;
+            synchronized (typesCollector)
+            {
+                if (typesCollector.stream().anyMatch(t -> Objects.equal(t.name, name) && t.formatClass == formatClass || t.formatClass.isAssignableFrom(formatClass) || formatClass.isAssignableFrom(t.formatClass)))
```
Isn't this expression always true for subtypes, e.g. the big format types?

I.e. is this missing a pair of brackets around the disjunction?

jacek-lewandowski replied:
Obviously `t.formatClass == formatClass` is redundant...

But to the point, I was thinking once about having something like intermediate abstract format, something like `SortedTableFormat` (and `SortedTableReader`) which would correspond to `SortedTableWriter`. Some components could be pushed down from the top level and I would use that extended verification as a sanity check.

blambov replied:
I mean that a subclass of `Component.Type` will always fire `t.formatClass.isAssignableFrom(formatClass)` (e.g. for `t = TOC`) and thus the whole condition without checking `Objects.equal(t.name, name)`.

blambov commented on src/java/org/apache/cassandra/io/sstable/SimpleGaugeProvider.java:
```diff
@@ -0,0 +1,72 @@
+
+    private static Iterable<SSTableReader> getAllReaders(Keyspace keyspace)
+    {
+        return Iterables.concat(Iterables.transform(keyspace.getColumnFamilyStores(), cfs -> cfs.getSSTables(SSTableSet.LIVE)));
```
Nit: for consistency, can this also use `cfs.getLiveSSTables()`?

blambov commented on src/java/org/apache/cassandra/io/sstable/BloomFilterMetrics.java:
```diff
@@ -0,0 +1,117 @@
+    /**
+     * False positive ratio of bloom filter from last read
+     */
+    public final GaugeProvider<Double> recentBloomFilterFalseRatio = newGaugeProvider("BloomFilterFalseRatio", readers -> {
```
This is the same as above, "Recent" needs to be added to the name as the calls below.

blambov commented on src/java/org/apache/cassandra/io/sstable/BloomFilterSupport.java:
```diff
@@ -0,0 +1,31 @@
+
+    BloomFilterTracker getBloomFilterTracker();
+
+    T cloneAndReplace(IFilter newBloomFilter);
```
What is the benefit of having this generic over simply making the implementations return the proper type?

Btw. this method appears to be unused.

jacek-lewandowski replied:
removed

blambov commented on src/java/org/apache/cassandra/io/sstable/format/big/BigTableReader.java:
```diff
@@ -687,4 +705,41 @@ private IndexSummary buildSummaryAtLevel(int newSamplingLevel) throws IOExceptio
+                if (cachedEntry != null)
+                {
+                    keyCacheHit.incrementAndGet();
+                    bloomFilterTracker.addTruePositive();
```
Preexisting problem, but this violates encapsulation and is much more naturally applied by the caller `getRowIndexEntry`.

blambov commented on src/java/org/apache/cassandra/io/sstable/KeyCacheSupport.java:
```diff
@@ -0,0 +1,41 @@
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.util.DataInputPlus;
+
+public interface KeyCacheSupport<T extends SSTableReader & KeyCacheSupport<T>>
```
`T` is never referenced, this should not be generic.

jacek-lewandowski replied:
it is with the recent changes, if we want to keep them

blambov commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -352,6 +354,8 @@ public MemtableOptions()
 
     public String[] data_file_directories = new String[0];
 
+    public Map<String, SSTableFormatConfig> sstable_formats = ImmutableMap.of("big", new SSTableFormatConfig(BigFormat.class.getName(), Collections.emptyMap()));
```
We usually use `ParameterizedClass` for class-with-options settings such as this one. Is it not suitable here?

jacek-lewandowski replied:
I didn't know that, thanks

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableFormat.java:
```diff
@@ -115,38 +123,75 @@ default boolean delete(Descriptor desc)
+            if (currentType != null)
+                return currentType;
+
+            String key = CassandraRelevantProperties.SSTABLE_FORMAT_DEFAULT.getString();
```
We should use the first defined format as the default if this option is not supplied.

jacek-lewandowski replied:
https://github.com/apache/cassandra/pull/2064/commits/80ef51bf6a43b9925949fa9e74c384dae0a5d6e9

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SSTableFormat.java:
```diff
@@ -115,38 +123,75 @@ default boolean delete(Descriptor desc)
-                if (valid.name.equalsIgnoreCase(name))
-                    return valid;
-            }
+            for (int i = 0; i < types.size(); i++)
```
Nit: let's just lowercase the name instead of using `equalsIgnoreCase` repeatedly.

jacek-lewandowski replied:
ok

jacek-lewandowski replied:
I just made the validation when initializing the configuration so that non-lowercased names are not accepted

jacek-lewandowski replied:
https://github.com/apache/cassandra/pull/2064/commits/80ef51bf6a43b9925949fa9e74c384dae0a5d6e9

blambov commented on src/java/org/apache/cassandra/service/CacheService.java:
```diff
@@ -467,9 +467,15 @@ public Future<Pair<KeyCacheKey, AbstractRowIndexEntry>> deserialize(DataInputPlu
+            SSTableFormat.Type type;
+            try
+            {
+                type = SSTableFormat.Type.getByOrdinal(typeOrdinal);
```
Some user will change the configuration in a way that is incompatible with this (e.g. reorder); we should put the identifier here.

jacek-lewandowski replied:
I'll add the identifier as a part of configuration in cassandra.yaml

jacek-lewandowski replied:
https://github.com/apache/cassandra/pull/2064/commits/80ef51bf6a43b9925949fa9e74c384dae0a5d6e9

blambov reviewed (commented):
Looks very good.

If it's not much work, it would be nice to have an abstract base class extending `SSTableReader` and adding support for the bloom filter (and possibly another one on top of it adding key cache).

blambov commented on src/java/org/apache/cassandra/io/sstable/keycache/KeyCacheSupport.java:
```diff
@@ -0,0 +1,120 @@
+    /**
+     * @return Number of key cache request
+     */
+    default long getKeyCacheRequest()
```
Nit: Request -> Requests

blambov commented on src/java/org/apache/cassandra/io/sstable/BloomFilterMetrics.java:
```diff
@@ -89,7 +89,7 @@ protected R map(SSTableReader r)
+    public final GaugeProvider<Double> recentBloomFilterFalseRatio = newGaugeProvider("RecentBloomFilterFalseRatio", readers -> {
         long falsePositiveCount = 0L;
         long truePositiveCount = 0L;
         long trueNegativeCount = 0L;
```
The calls below (`getFalsePositiveCount` etc) should be changed to their recent versions.

blambov commented on src/java/org/apache/cassandra/io/sstable/format/SortedTableReader.java:
```diff
@@ -0,0 +1,120 @@
+
+import static org.apache.cassandra.utils.concurrent.SharedCloseable.sharedCopyOrNull;
+
+public abstract class SortedTableReader extends SSTableReader
```
Nit: Because the only thing this provides is bloom filter support, doesn't it make better sense to call it `SSTableReaderWithFilter`?

blambov reviewed (approved):
Latest changes look good, and I believe they complete my huge range of requests. Thank you for the patience with them.

blambov commented on test/unit/org/apache/cassandra/Util.java:
```diff
@@ -1236,4 +1241,18 @@ public static void flush(TableViews view)
+    public static DataInputPlus.DataInputStreamPlus createDataInputStreamPlus(InputStream in)
+    {
+        DataInputStreamPlusImpl impl = new DataInputStreamPlusImpl(in);
+        return Mockito.mock(DataInputPlus.DataInputStreamPlus.class, new ForwardsInvocations(impl));
```
Isn't it risky to always mock (i.e. track invocations), which can cause a lot of overhead?

jacek-lewandowski replied:
Just very few unit tests uses that in a very inexcessive way

jacek-lewandowski committed:
SSTable format API

Summary of the changes:

Format, reader and writer
---------------------------
There are a lot of refactorings around sstable related classes aiming to extract the most generic functionality to the top-level entities and push down implementation-specific stuff to the actual implementation. In Particular, the top-level, implementation agnostic classes/interfaces are SSTableFormat interface, SSTable, SSTableReader, SSTableWriter, IVerifier, and IScrubber. The rest of the codebase has been reviewed for explicit usages of big table format-specific usages of sstable classes and refactored. SSTable, SSTableReader, and SSTableWriter have their builders. Builders make a hierarchy that follows the same inheritance structure as readers and writers.

There are also partial implementations that add support for some features and may or may not be used by the custom implementations. They include:
- AbstractSSTableFormat - adds an implementation of some initialization methods - in practice, all of the format implementations should extend this class
- SSTableReaderWithFilter - add support for Bloom filter to the reader
- SortedTableWriter - generic implementation for a writer which writes partitions in the default order to the data file, supports Bloom filter and some index of partitions
- IndexSummarySupport - interface implemented by the readers using index summaries
- KeyCacheSupport - interface implemented by the readers using row key cache

Descriptor
---------------------------
Refactored the Descriptor class so that:
- All paths are created from the base directory File rather than from a String
- All the methods named *filename* producing full paths were made private; their current implementations are returning file names rather than paths (the naming was inconsistent)
- The usages of the `filenameFor` method were refactored to use the `fileFor` method
- The usages of the `fromFilename` method were refactored to use a  `fromFileWithComponent(..., false).left` expression
In essence, the Descriptor class is no longer working on String-based paths.

Index summaries
---------------------------
Removed the index summary from the generic SSTableReader class and created an interface IndexSummarySupport to be implemented by the readers that need it. Methods in related classes that refer back to the reader were refactored to support just readers of the SSTableReader & IndexSummarySupport type. Therefore, we will no longer need to assume that the generic SSTableReader has anything to do with an index summary.

A new IndexSummaryComponent class encloses data fields from the index summary file (note that aside from the index summary itself, the file includes the first and last partition of the sstable). The class has been extracted to deal with those fields and have that logic in a single place.

Filter
---------------------------
Refactored IFilter and its serialization - in particular, added the `serialize` method to the IFilter interface and moved loading/saving logic to a separate utility class FilterComponent.
Extracted the SSTableReaderWithFilter abstract reader extending the generic SSTableReader with filter support.
Extracted bloom filter metrics into separate entities allowing to plug them in if the implementation uses a filter.

Cache
---------------------------
Refactored CacheService to support different key-cache values. CacheService now supports arbitrary IRowIndexEntry implementation as a key-cache value. A new version of the auto-saving cache was created ("g") because some information about the type of serialized row index entry needs to be known before it is deserialized (or skipped). Therefore, the SSTableFormat type ordinal number is stored, which is sufficient because the IRowIndexEntry serializer is specific to the sstable format type.
Similarly to the IndexSummarySupport, a new KeyCacheSupport interface has to be implemented to mark the reader as supporting key-cache. It contains the default implementation of several methods the rest of the system relies on when the key-cache is supported.

Other changes
---------------------------
- Fixed disabling chunk cache - enable(boolean) method in ChunkCache does not make any sense - it makes a false impression it can disable chunk cache once enabled, while in fact, it only clears it. Added setFileCacheEnabled to DatabaseDescriptor

- Made WrappingUnfilteredRowIterator an interface

- DataInputStreamPlus extends InputStream - this makes it possible for input stream-based inheritors of DataInputPlus to extend DataInputStreamPlus. It simplifies coding because sometimes we want to get DataInputPlus implementation extending InputStream as an argument.

- Table and keyspace metrics were made pluggable - in particular, added the ability for a certain format to register gauges that are specific only to that format and make no sense for others

- Implemented mmapped region extension for compressed data

- Refactored FileHandle so that it is no longer closable

- Implemented WrappingRebufferer

- Introduced the SSTable.Owner interface to make SSTable implementation not reference higher-level entities directly. SSTable accepts passing null as the owner when there is no owner (like sometimes in offline tools) or passing a mock when needed in tests.

Individual commits
---------------------------

[4a87cd36fe] Fix disabling chunk cache
[c84c75ccf3] Made WrappingUnfilteredRowIterator an interface
[253d2b828e] Add getType to SSTableFormat
[3f169dcc20] Remove getIndexSerializer from SSTableFormat
[05bae1833b] Pull down rowIndexEntrySerializer field
[da675f2809] Moved RowIndexEntry
[673f0c5c39] Reduce usages of RowIndexEntry
[c72538be91] Refactor CacheService to support for different key cache values
[54d33ee656] Minor refactoring of ColumnIndex
[93862df967] Just moved AbstractSSTableIterator to o.a.c.io.sstable.format
[9e4566a1de] Refactored AbstractSSTableIterator
[a4e61e80bb] Extracted IScrubber and IVerifier interfaces
[20f78c7419] Push down implementation of SSTableReader.firstKeyBeyond
[f2c24e5774] Moved SSTableReader.getSampleIndexesForRanges to IndexSummary
[b6c3a6c1ea] Moved SSTableReader.getKeySamples implementation to IndexSummary
[c4b90ebb33] Refactor InstanceTidier so that it is more generic
[918d5a9e74] Refactor dropping page cache
[a52fb4d558] Refactor sstable metrics
[f6d10f930f] NEW (fix up) - DataInputStreamPlus extends InputStream
[8f6a56d972] Getting rid of index summary in SSTableReader
[4a918bf725] Removed direct usages of primary index from SSTableReader
[358fa32602] Refactor KeyIterator so that it is sstable format agnostic
[14c09d89c2] Remove explicit usage of Components outside of format specific classes
[feff14e137] Move clone methods implementation from SSTableReader to BigTableReader
[64e9787b10] Move saveIndexSummary and saveBloomFilter to SSTableReaderBuilder
[ae71fe6ed8] Moved indexSummary field to BigTableReader and made it private
[df9fd8c4b9] Moved ifile field to BigTableReader and made it private
[2be6ea9ecf] Moved static open methods for BigTableReader to the reader factory
[bc0e55ac48] Minor refactoring around IFilter and its serialization
[5b95704beb] Minor refactorings around IndexSummary
[87812335e8] Extracted TOCComponent class to deal with TOC file
[fdad092a6a] Extracted CompressionInfoComponent class
[39b47e388d] Extracted StatsComponent as a helper for elements of SSTable metadata
[cdb55bff47] Fix SSTable.getMinimalKey
[b99c6d5805] Refactor FileHandle so that it is no longer closable
[77b7f7ace5] Implement WrappingRebufferer
[b6868914dd] Add progressPercentage to ProgressInfo
[7fd4956e5b] Moved copy/rename/hardLink methods from SSTableWriter to SSTable
[1ccc6bf148] Create generic SSTableBuilder and IOOptions
[da58a81102] Refactor SSTableReaderBuilder
[4501ddba1c] Refactor ColumnIndex
[d4f9e1a64b] Extracted non-big-table-specific functionality from BigTableWriter to SortedTableWriter
[379525d01e] Refactor BigTableZeroCopyWriter to SSTableZeroCopyWriter as it is not specific to big format
[8ac37f83bc] Extract EmptySSTableScanner out from BigTableScanner
[ee6673f1cf] Implement SSTableWriterBuilder
[bb26629235] Refactor opening early / final
[a327595015] Refactored SSTableWriter factory
[16ffd7334b] Extract non-big-format-specific logic from scrubber and verifier
[75e02db6af] Allow to specify the default SSTableFormat via system property
[a7b9d0d628] Small fixes around streaming
[407f977c36] Move guard collection size
[0529e57d2f] Remove explicit references to big format
[61509963ec] Unclassified minor changes
[da28d1af3a] Replaced getCreationTimeFor(Component) with getDataCreationTime()
[e99c834de6] !!! Reformatting
[882b7baa5a] Rename SSTableReader.maybePresent and fix its redundant usages
[b70c983bea] Implement mmapped region extension for compressed data
[d7ff3970de] Introduce SSTable.Owner interface
[e9feb9c462] Replaced getCreationTimeFor(Component) with getDataCreationTime()
[ee8082fb07] Created SSTableFormat.deleteOrphanedComponents
[e62950fd3d] Refactor metrics further
[cefa5b3814] Extract key cache support into separate entity
[dd55101ca1] Extracted SSTableReaderWithFilter
[510b651824] Implement customizable component types
[2be512d9fa] Pluggable SSTableFormat by making SSTableFormat.Type not an enum
[670836b55d] Refactor CRC and digest validators
[00c91103bc] Extract delete method to delete SSTables and purge row cache entries
[0819dc9fc2] Extracted trySkipFileCacheBefore(key) to SSTableReader
[732f841750] Added missing overrides in ForwardingSSTableReader
[db623218fd] Update DatabaseDescriptorRefTest
[c018c468e5] Cleanup
[eafc836242] Add @SuppressWarnings("resource") where needed
[3b7c911dd6] Documentation

patch by Jacek Lewandowski, reviewed by Branimir Lambov for CASSANDRA-17056

Co-authored-by: @jacek-lewandowski
Co-authored-by: @blambov
